File: entities/Spell.py
from entities.Area import *
from entities.DifficultyClass import *
from database import *

class SpellDamage:
    def __init__(self, damage):
        self.damageType = damage["damage_type"]["index"] if "damage_type" in damage else None
        self.damageSlots = damage["damage_at_slot_level"] if "damage_at_slot_level" in damage else None
        self.damageAtCharacterLevel = damage["damage_at_character_level"] if "damage_at_character_level" in damage else None

class Spell:
    def __init__(self, spell):
        self.name = spell["index"]
        self.range = spell["range"]
        self.ritual = spell["ritual"]
        self.duration = spell["duration"]
        self.concentration = spell["concentration"]
        self.castingTime = spell["casting_time"]
        self.healAtSlotLevel = spell["heal_at_slot_level"] if "heal_at_slot_level" in spell else None
        self.school = spell["school"]["index"]
        self.classes = [item["index"] for item in spell["classes"]]
        self.subclasses = [item["index"] for item in spell["subclasses"]] if "subclasses" in spell else None
        self.areaEffect = Area(spell["area_of_effect"]) if "area_of_effect" in spell else None
        self.dc = DifficultyClass(spell["dc"]) if "dc" in spell else None
        self.damage = SpellDamage(spell["damage"]) if "damage" in spell else None
        self.attackType = spell["attack_type"] if "attack_type" in spell else None

spellStats = [Spell(item) for item in list(Database.getAllSpells(db))]

File: TrainDataGen/Entities/Spell.cs
using TrainDataGen.Entities.Enums;
using TrainDataGen.Entities.Mappers;
using TrainDataGen.Utilities;

namespace TrainDataGen.Entities;

public class Spell: BaseEntity
{
    public string Range { get; set; }
    public byte Level { get; set; }
    public string Duration { get; set; }
    public bool Concentration { get; set; }
    public string CastingTime { get; set; }
    public Dictionary<string, string>? HealAtSlotLevel { get; set; }
    public BaseEntity School { get; set; }
    public List<BaseEntity> Classes { get; set; }
    public List<BaseEntity>? Subclasses { get; set; }
    public Area? AreaEffect { get; set; }
    public DifficultyClass? Dc { get; set; }
    public SpellDamage? Damage { get; set; }
    public string? AttackType { get; set; }

    public Spell(SpellMapper spell): base(spell.Index, spell.Name)
    {
        Index = spell.Index;
        Name = spell.Name;
        Range = spell.Range;
        Level = spell.Level;
        Duration = spell.Duration;
        Concentration = spell.Concentration;
        CastingTime = spell.CastingTime;
        HealAtSlotLevel = spell.HealAtSlotLevel;
        School = spell.School;
        Classes = spell.Classes;
        Subclasses = spell.Subclasses;
        AreaEffect = spell.AreaOfEffect;
        Dc = spell.Dc;
        Damage = spell.Damage != null ? new SpellDamage(spell.Damage) : null;
        AttackType = spell.AttackType;
    }
}

public class SpellDamage
{
    public string? DamageType { get; set; }
    public Dictionary<string, string>? DamageSlots { get; set; }
    public Dictionary<string, string>? DamageAtCharacterLevel { get; set; }

    public SpellDamage(dynamic damage)
    {
        DamageType = damage.damage_type != null ? damage.damage_type.index : null;
        DamageSlots = damage.damage_at_slot_level != null ? damage.damage_at_slot_level as Dictionary<string, string> : null;
        DamageAtCharacterLevel = damage.damage_at_character_level != null ? damage.damage_at_character_level as Dictionary<string, string> : null;
    }
}